folding-schemes/src/folding/hypernova/decider_eth.rs

```diff
@@ -0,0 +1,309 @@
+    FS: FoldingScheme<C1, C2, FC>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'b> &'b GC1: GroupOpsBounds<'b, C1, GC1>,
+    for<'b> &'b GC2: GroupOpsBounds<'b, C2, GC2>,
+    // constrain FS into HyperNova, since this is a Decider specifically for HyperNova
+    HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU, false>: From<FS>,
+    crate::folding::hypernova::ProverParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::ProverParam>,
+    crate::folding::hypernova::VerifierParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::VerifierParam>,
+{
+    type PreprocessorParam = (FS::ProverParam, FS::VerifierParam);
+    type ProverParam = (S::ProvingKey, CS1::ProverParams);
+    type Proof = Proof<C1, CS1, S>;
+    /// VerifierParam = (pp_hash, snark::vk, commitment_scheme::vk)
+    type VerifierParam = (C1::ScalarField, S::VerifyingKey, CS1::VerifierParams);
+    type PublicInput = Vec<C1::ScalarField>;
+    type CommittedInstance = ();
+
+    fn preprocess(
+        mut rng: impl RngCore + CryptoRng,
+        prep_param: &Self::PreprocessorParam,
+        fs: FS,
+    ) -> Result<(Self::ProverParam, Self::VerifierParam), Error> {
+        let circuit =
+            DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+                fs.into(),
+            )
+            .unwrap();
+
+        // get the Groth16 specific setup for the circuit
+        let (g16_pk, g16_vk) = S::circuit_specific_setup(circuit, &mut rng).unwrap();
+
+        // get the FoldingScheme prover & verifier params from HyperNova
+        #[allow(clippy::type_complexity)]
+        let hypernova_pp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::ProverParam = prep_param.0.clone().into();
+        #[allow(clippy::type_complexity)]
+        let hypernova_vp: <HyperNova<C1, GC1, C2, GC2, FC, CS1, CS2, MU, NU,false> as FoldingScheme<
+            C1,
+            C2,
+            FC,
+        >>::VerifierParam = prep_param.1.clone().into();
+        let pp_hash = hypernova_vp.pp_hash()?;
+
+        let pp = (g16_pk, hypernova_pp.cs_pp);
+        let vp = (pp_hash, g16_vk, hypernova_vp.cs_vp);
+        Ok((pp, vp))
+    }
+
+    fn prove(
+        mut rng: impl RngCore + CryptoRng,
+        pp: Self::ProverParam,
+        folding_scheme: FS,
+    ) -> Result<Self::Proof, Error> {
+        let (snark_pk, cs_pk): (S::ProvingKey, CS1::ProverParams) = pp;
+
+        let circuit = DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+            folding_scheme.into(),
+        )?;
+
+        let snark_proof = S::prove(&snark_pk, circuit.clone(), &mut rng)
```

Comment:
Is there any reason why `S::prove` forces to pass the circuit itself and not a reference to it? We could avoid costly clones on this way if we have decently big circuits.

Reply:
That comes from arkworks: 
https://github.com/arkworks-rs/snark/tree/202b9e02b808d1d15823298b8ac16f5c54ae4ef6/snark/src/lib.rs#L50

Reply:
Will file an issue for this.